main.lua

```diff
@@ -1234,22 +1251,103 @@ end
+    end
+
+    local newBuffer = buffer.NewBuffer(bufContents, newBufferTitle)
+    newBuffer.Type.Scratch = true
+    newBuffer.Type.Readonly = true
+    micro.CurPane():HSplitBuf(newBuffer)
+end
+
+-- takes Location[] https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#location
+-- and renders them to user
+function showReferenceLocations(newBufferTitle, lspLocations)
+    local references = {}
     for i, lspLoc in ipairs(lspLocations) do
         local fpath = absPathFromFileUri(lspLoc.uri)
         local lineNumber = lspLoc.range.start.line + 1
         local columnNumber = lspLoc.range.start.character + 1
-        local line = string.format("%s:%d:%d\n", fpath, lineNumber, columnNumber)
-        if labels ~= nil then
-            line = labels[i] .. "\t" .. line
+        references[i] = {
+            path = fpath,
+            line = lineNumber,
+            column = columnNumber,
+        }
+    end
+
+    table.sort(references, function(a, b)
+        if a.path < b.path then return true
+        elseif a.path > b.path then return false
+        else
+            if a.line < b.line then return true
+            elseif a.line > b.line then return false
+            else
+                if a.column < b.column then return true
+                elseif a.column > b.column then return false
+                else return true end
+            end
         end
-        bufContents = bufContents .. line
+    end)
+
+    local bufContents = ""
+    local curFile = nil
+    local file = nil
+    local lineCount = 0
+    local prevLine = 0
+    local lineContent
+    for _, ref in ipairs(references) do
+        repeat -- fake-goto to use break as continue
+            if curFile == ref.path and prevLine == ref.line then break -- Same line? add it directly
+            elseif curFile ~= ref.path then --Handle file opening and closing
+                if file then file:close() end
+                curFile = ref.path
+                if bufContents ~= "" then bufContents = bufContents .. "\n" end
+                bufContents = bufContents .. curFile .. ":\n"
+                lineCount = 0
+                file = io.open(curFile, "rb")
+                if not file then
+                    lineContent = ""
+                    break
+                end
+            end
+
+            repeat
+                lineContent = file:read("*l")
+                lineCount = lineCount + 1
+            until not lineContent or lineCount == ref.line
+            ---Empty string: line not found
+            lineContent = lineContent and lineContent or ""
+
+        until true
+        bufContents = bufContents .. string.format("\t%d:%d:%s\n",ref.line, ref.column, lineContent)
+        prevLine = ref.line
     end
 
+    if file then file:close() end -- last iteration does not close last file
+
     local newBuffer = buffer.NewBuffer(bufContents, newBufferTitle)
```

Comment:
I think this is too complex, it's hard to follow so deeply nested logic. This suggestion also changes it to collect the lines into an array to avoid repeated string concatenation, which improves readability and should be way more efficient too when there are lots of references.
```suggestion
    local bufLines = {}
    local curFilePath = ""
    local file = nil
    local lineCount = 0
    local lineContent = ""
    for _, ref in ipairs(references) do
        if curFilePath ~= ref.path then
            if file then file:close() end
            if #bufLines > 0 then table.insert(bufLines, "") end
            curFilePath = ref.path
            table.insert(bufLines, curFilePath)
            file = io.open(curFilePath, "rb")
            lineCount = 0
        end

        -- file can be nil if io.open failed
        if file ~= nil then
            while lineCount < ref.line do
                lineContent = file:read("*l")
                lineCount = lineCount + 1
            end
        end
        table.insert(bufLines, string.format("\t%d:%d:%s", ref.line, ref.column, lineContent or ""))
    end

    if file then file:close() end -- last iteration does not close last file
    table.insert(bufLines, "")

    local newBuffer = buffer.NewBuffer(table.concat(bufLines, "\n"), newBufferTitle)
```